File: src/19p2.py
import functools


@functools.cache
def num_possibilities(pattern: str) -> int:
    prefixes = [d for d in designs if pattern.startswith(d)]
    cnt = 0
    for p in prefixes:
        cnt += 1 if pattern == p else num_possibilities(pattern.removeprefix(p))
    return cnt


if __name__ == "__main__":
    with open("../inp/19.txt", 'r') as f:
        lines = f.readlines()

    designs = set(lines[0].strip().split(", "))
    patterns = set(l.strip() for l in lines[2:])

    out = 0
    for pat in patterns:
        out += num_possibilities(pat)
    print(f"Final result: {out}")

File: src/19p2.kt
fun main() {
    val input = readInput("19")
    val designs = input[0].split(", ").toSet()
    val patterns = buildSet {
        input.subList(2, input.size).forEach { line -> this.add(line) }
    }

    val cache = mutableMapOf<String, Long>()
    fun numPossibilities(pattern: String): Long = cache.getOrPut(pattern) {
        val prefixes = designs.filter { pattern.startsWith(it) }
        if (prefixes.isEmpty()) {
            0L
        } else {
            var out = 0L
            for (pr in prefixes) {
                out += if (pattern == pr) {
                    1L
                } else {
                    numPossibilities(pattern.removePrefix(pr))
                }
            }
            out
        }
    }

    patterns.sumOf { numPossibilities(it) }.println()
}
